execbuilder: elide redundant locking lookup join in some simple cases

michae2 opened the pull request:
**execbuilder: elide redundant locking lookup join in some simple cases**

In some simple cases of SELECT FOR UPDATE using the new Lock operator, we can elide the locking lookup join and instead push the locking down into the input.

To optimize more complex cases we'll need to turn locking into a physical property which is required by the new Lock operator, similar to ordering.

Fixes: #114566

Release note: None

---

**opt: add read committed variant of tpcc opttest**

Add a copy of the tpcc opttest, but with read committed isolation. This shows the locking differences between plans built for serializable and plans built for read committed.

This doesn't show the optimization in the previous commit, because it applies during execbuilder.

Release note: None

michae2 committed:
execbuilder: elide redundant locking lookup join in some simple cases

In some simple cases of SELECT FOR UPDATE using the new Lock operator,
we can elide the locking lookup join and instead push the locking down
into the input.

To optimize more complex cases we'll need to turn locking into a
physical property which is required by the new Lock operator, similar to
ordering.

Fixes: #114566

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/114401)
<!-- Reviewable:end -->

michae2 committed:
opt: add read committed variant of tpcc opttest

Add a copy of the tpcc opttest, but with read committed isolation. This
shows the locking differences between plans built for serializable and
plans built for read committed.

This doesn't show the optimization in the previous commit, because it
applies during execbuilder.

Release note: None

michae2 commented:
I wanted to show a benchmark, but was failing to get the TPC-C roachtest running last night from my gceworker. I'll try to do that again sometime today.

michae2 commented:
@DrewKimball asked whether this should be an optimization rule, and also whether we would like more optimization rules for SFU, to allow the optimizer to choose plans that don't have the lookup join. That would allow us to avoid the lookup join in more cases.

I agree that an optimization rule would be more optimal. We can't use an optimization rule that removes the `Lock` operator, as we need that to act as an optimization barrier to prevent removing the locking side effect. Drew suggested an optimization rule that changed `Lock`'s locking to ForNone. I am considering this.

If it helps, here was my thinking for optimization of SFU:
- Backport this PR to 23.2.
- Revert this PR in 24.1 and instead make locking a physical property that is required by the Lock operator and provided by scan, index join, and lookup join, with lookup join as an enforcer. I think this will allow the optimizer to pick plans without the lookup join whenever its optimal, while keeping Lock as an optimization barrier.

DrewKimball reviewed (commented):
> Reviewed 3 of 3 files at r1, 3 of 3 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114401)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner, @michae2, and @nvanbenschoten)

___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1173 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114401#-NjP73Nf5LborIu7vbcC:-NjP73Nf5LborIu7vbcD:b-jghoax) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c06bdef879e19a36165dd427cdecfdfb759100d/pkg/sql/opt/exec/execbuilder/mutation.go#L1173)):*
> ```Go
> 	case *memo.LookupJoinExpr:
> 		if input.Table == lock.Table && input.Index == cat.PrimaryIndex &&
> 			input.JoinType == opt.InnerJoinOp && len(input.On) == 0 {
> ```

I think the tableID check ensures that the key columns are the same in practice, but it would be best to verify that.

We should probably check `IsFirstJoinInPairedJoiner` and `IsSecondJoinInPairedJoiner` as well.

Also, I'm assuming it's ok to lock the same row twice? There's no guarantee here and for the index-join case that the input doesn't join to the same row twice.
___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1173 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114401#-NjP8r0q638PQ8YPyq87:-NjP8r0q638PQ8YPyq88:byib6tr) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c06bdef879e19a36165dd427cdecfdfb759100d/pkg/sql/opt/exec/execbuilder/mutation.go#L1173)):*
> ```Go
> 	case *memo.LookupJoinExpr:
> 		if input.Table == lock.Table && input.Index == cat.PrimaryIndex &&
> 			input.JoinType == opt.InnerJoinOp && len(input.On) == 0 {
> ```

We could probably also allow semi-join and left-join here.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/select_for_update` line 2752 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114401#-NjP892ICSadSIHEJbC3:-NjP892ICSadSIHEJbC4:b-6gk8vc) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c06bdef879e19a36165dd427cdecfdfb759100d/pkg/sql/opt/exec/execbuilder/testdata/select_for_update#L2752)):*
> ```
> # Push the locking down into the lookup join.
> query T
> EXPLAIN (VERBOSE) SELECT * FROM t JOIN u ON u.a = t.b WHERE t.a = 5 FOR UPDATE OF u
> ```

Another test idea: a lookup join where the input has duplicates, so the same row gets looked up more than once.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/select_for_update` line 2800 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114401#-NjP8NFGErd10REMhO9F:-NjP8NFGErd10REMhO9G:b7xlb97) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c06bdef879e19a36165dd427cdecfdfb759100d/pkg/sql/opt/exec/execbuilder/testdata/select_for_update#L2800)):*
> ```
>           table: t@t_pkey
>           spans: /5/0
> ```

Can we also test the following?
1. Scan/lookup-join on a non-primary index.
2. Scan/index-join/lookup-join on a different table.
3. There is a lookup-join between a table and itself.
   a. The right side of the join is locked.
   b. The left side of the join is locked.


<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
Thanks for the comments!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114401#-:-NjfabS60SfXsX-n_FvQ:b-a09w3v)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @mgartner, and @nvanbenschoten)

___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1173 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114401#-NjP73Nf5LborIu7vbcC:-NjfOi67Bo8qvoo4FFTc:b2ulc2j) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c06bdef879e19a36165dd427cdecfdfb759100d/pkg/sql/opt/exec/execbuilder/mutation.go#L1173)):*
> I think the tableID check ensures that the key columns are the same in practice, but it would be best to verify that.

Maybe I'm misunderstanding, but I don't think we need the key columns to be the same. In fact, I wouldn't expect them to ever be the same due to the tableID check. For example, suppose we had a query like the following:

```sql
CREATE TABLE ab (a INT PRIMARY KEY, b INT);
CREATE TABLE cde (c INT, d INT, e INT, PRIMARY KEY (c, d));
SET optimizer_use_lock_op_for_serializable = true;
SELECT * FROM ab JOIN cde ON c = b WHERE a = 1 FOR UPDATE OF cde;
```

Here's what the explain looks like before the optimization:

```
demo@127.0.0.1:26257/demoapp/defaultdb> EXPLAIN (OPT, VERBOSE) SELECT * FROM ab JOIN cde ON c = b WHERE a = 1 FOR UPDATE OF cde;
                                              info
------------------------------------------------------------------------------------------------
  lock cde
   ├── columns: a:1 b:2 c:5 d:6 e:7
   ├── locking: for-update
   ├── volatile, mutations
   ├── stats: [rows=9.9]
   ├── cost: 54.217
   ├── key: (6)
   ├── fd: ()-->(1,2,5), (6)-->(7), (2)==(5), (5)==(2)
   ├── distribution: us-east1
   └── inner-join (lookup cde)
        ├── columns: a:1 b:2 c:5 d:6 e:7
        ├── key columns: [2] = [5]
        ├── stats: [rows=9.9, distinct(2)=1e-10, null(2)=0, distinct(5)=1e-10, null(5)=0]
        ├── cost: 54.197
        ├── key: (6)
        ├── fd: ()-->(1,2,5), (6)-->(7), (2)==(5), (5)==(2)
        ├── distribution: us-east1
        ├── scan ab
        │    ├── columns: a:1 b:2
        │    ├── constraint: /1: [/1 - /1]
        │    ├── cardinality: [0 - 1]
        │    ├── stats: [rows=1, distinct(1)=1, null(1)=0, distinct(2)=0.995512, null(2)=0.01]
        │    ├── cost: 9.09
        │    ├── key: ()
        │    ├── fd: ()-->(1,2)
        │    ├── distribution: us-east1
        │    └── prune: (2)
        └── filters (true)
(28 rows)
```

In this case the key columns of the inner lookup join are [2] and the key columns of the locking semi lookup join are [5, 6] (not printed) (which originate from the inner lookup join, ensured by the tableID check). We don't need the key columns to be the same to know that the locking lookup join will lock the same set of PKs as produced by the inner lookup join. We just need the inner lookup join to be on the same primary index, and not have ON conditions, and not be an anti-join.

> We should probably check `IsFirstJoinInPairedJoiner` and `IsSecondJoinInPairedJoiner` as well.

Done.

> Also, I'm assuming it's ok to lock the same row twice?

Yes (locking the same row twice was possible before the optimization, too). I've added a case to `pkg/sql/logictest/testdata/logic_test/select_for_update_read_committed` that does this.
___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1173 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114401#-NjP8r0q638PQ8YPyq87:-NjfPMS86oUOycl8Wfck:b-4tyov) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c06bdef879e19a36165dd427cdecfdfb759100d/pkg/sql/opt/exec/execbuilder/mutation.go#L1173)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

We could probably also allow semi-join and left-join here.
</blockquote></details>

Good point. Added semi (although I'm having trouble producing it due to https://github.com/cockroachdb/cockroach/issues/114712).

Won't left joins be disallowed when I fix https://github.com/cockroachdb/cockroach/issues/97434?
___
*[`pkg/sql/opt/exec/execbuilder/testdata/select_for_update` line 2752 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114401#-NjP892ICSadSIHEJbC3:-NjfPgfg8QxNGOCfvxvh:b-ka198r) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c06bdef879e19a36165dd427cdecfdfb759100d/pkg/sql/opt/exec/execbuilder/testdata/select_for_update#L2752)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Another test idea: a lookup join where the input has duplicates, so the same row gets looked up more than once.
</blockquote></details>

Done (in `pkg/sql/logictest/testdata/logic_test/select_for_update_read_committed`).
___
*[`pkg/sql/opt/exec/execbuilder/testdata/select_for_update` line 2800 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114401#-NjP8NFGErd10REMhO9F:-NjfPoKn8nIdUmqUMG-U:b-p9ff45) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c06bdef879e19a36165dd427cdecfdfb759100d/pkg/sql/opt/exec/execbuilder/testdata/select_for_update#L2800)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Can we also test the following?
1. Scan/lookup-join on a non-primary index.
2. Scan/index-join/lookup-join on a different table.
3. There is a lookup-join between a table and itself.
   a. The right side of the join is locked.
   b. The left side of the join is locked.
</blockquote></details>

Did all except scan on a different table. (Is it possible to produce a plan with scans of two different tables without a join on top of the scans?)


<!-- Sent from Reviewable.io -->

mgartner reviewed (commented):
> Reviewed 1 of 3 files at r1, 6 of 6 files at r3, 1 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114401)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball and @nvanbenschoten)

___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1173 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114401#-NjP73Nf5LborIu7vbcC:-NjrkmKZ681k4nDzZ25c:b-n97ijv) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c06bdef879e19a36165dd427cdecfdfb759100d/pkg/sql/opt/exec/execbuilder/mutation.go#L1173)):*
<details><summary><i>Previously, michae2 (Michael Erickson) wrote…</i></summary><blockquote>

> I think the tableID check ensures that the key columns are the same in practice, but it would be best to verify that.

Maybe I'm misunderstanding, but I don't think we need the key columns to be the same. In fact, I wouldn't expect them to ever be the same due to the tableID check. For example, suppose we had a query like the following:

```sql
CREATE TABLE ab (a INT PRIMARY KEY, b INT);
CREATE TABLE cde (c INT, d INT, e INT, PRIMARY KEY (c, d));
SET optimizer_use_lock_op_for_serializable = true;
SELECT * FROM ab JOIN cde ON c = b WHERE a = 1 FOR UPDATE OF cde;
```

Here's what the explain looks like before the optimization:

```
demo@127.0.0.1:26257/demoapp/defaultdb> EXPLAIN (OPT, VERBOSE) SELECT * FROM ab JOIN cde ON c = b WHERE a = 1 FOR UPDATE OF cde;
                                              info
------------------------------------------------------------------------------------------------
  lock cde
   ├── columns: a:1 b:2 c:5 d:6 e:7
   ├── locking: for-update
   ├── volatile, mutations
   ├── stats: [rows=9.9]
   ├── cost: 54.217
   ├── key: (6)
   ├── fd: ()-->(1,2,5), (6)-->(7), (2)==(5), (5)==(2)
   ├── distribution: us-east1
   └── inner-join (lookup cde)
        ├── columns: a:1 b:2 c:5 d:6 e:7
        ├── key columns: [2] = [5]
        ├── stats: [rows=9.9, distinct(2)=1e-10, null(2)=0, distinct(5)=1e-10, null(5)=0]
        ├── cost: 54.197
        ├── key: (6)
        ├── fd: ()-->(1,2,5), (6)-->(7), (2)==(5), (5)==(2)
        ├── distribution: us-east1
        ├── scan ab
        │    ├── columns: a:1 b:2
        │    ├── constraint: /1: [/1 - /1]
        │    ├── cardinality: [0 - 1]
        │    ├── stats: [rows=1, distinct(1)=1, null(1)=0, distinct(2)=0.995512, null(2)=0.01]
        │    ├── cost: 9.09
        │    ├── key: ()
        │    ├── fd: ()-->(1,2)
        │    ├── distribution: us-east1
        │    └── prune: (2)
        └── filters (true)
(28 rows)
```

In this case the key columns of the inner lookup join are [2] and the key columns of the locking semi lookup join are [5, 6] (not printed) (which originate from the inner lookup join, ensured by the tableID check). We don't need the key columns to be the same to know that the locking lookup join will lock the same set of PKs as produced by the inner lookup join. We just need the inner lookup join to be on the same primary index, and not have ON conditions, and not be an anti-join.

> We should probably check `IsFirstJoinInPairedJoiner` and `IsSecondJoinInPairedJoiner` as well.

Done.

> Also, I'm assuming it's ok to lock the same row twice?

Yes (locking the same row twice was possible before the optimization, too). I've added a case to `pkg/sql/logictest/testdata/logic_test/select_for_update_read_committed` that does this.
</blockquote></details>

Is it safe to push the locks into locality-optimized lookup joins (e.g., when `RemoteLookupExpr FiltersExpr` is non-nil, or when `LocalityOptimized bool` is true)? I believe we are free to not lock rows if we can answer the query without scanning them, so I think not locking remote rows in a LOS when we find all the rows locally should be ok, correct?


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> Reviewed 6 of 6 files at r3, 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114401)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @nvanbenschoten)

___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1173 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114401#-NjP8r0q638PQ8YPyq87:-NjsLTeZD0H3E5vXdLMI:b-p7qp1g) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c06bdef879e19a36165dd427cdecfdfb759100d/pkg/sql/opt/exec/execbuilder/mutation.go#L1173)):*
>Won't left joins be disallowed when I fix [![](https://avatars.githubusercontent.com/u/46886010?s=60&v=4) opt: don't allow locking null-extended rows #97434](https://github.com/cockroachdb/cockroach/issues/97434)?

Good point, maybe we shouldn't worry about that then.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/select_for_update` line 2800 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114401#-NjP8NFGErd10REMhO9F:-NjsMNOdEXvF6XiA2PEG:btbhcrq) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c06bdef879e19a36165dd427cdecfdfb759100d/pkg/sql/opt/exec/execbuilder/testdata/select_for_update#L2800)):*
>Did all except scan on a different table. (Is it possible to produce a plan with scans of two different tables without a join on top of the scans?)

Thanks. Yeah, I don't think that's actually possible, good point.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/select_for_update` line 2773 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114401#-NjsNWvt2LyWSqA8wMSn:-NjsNWvt2LyWSqA8wMSo:b74abk6) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0ea8ec5414592832febfcabb156ca11ef34d251/pkg/sql/opt/exec/execbuilder/testdata/select_for_update#L2773)):*
> ```
>       spans: /5-/6
> 
> # Index join on a different table. In this case we can't push the locking down.
> ```

Looks like this one doesn't end up as an index-join.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114401)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @nvanbenschoten)

___
*[`pkg/sql/opt/exec/execbuilder/testdata/select_for_update` line 2773 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114401#-NjsNWvt2LyWSqA8wMSn:-NkHARpM-iU15rPZsBHb:buwa3c6) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0ea8ec5414592832febfcabb156ca11ef34d251/pkg/sql/opt/exec/execbuilder/testdata/select_for_update#L2773)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Looks like this one doesn't end up as an index-join.
</blockquote></details>

Hinting the join should make this easier, I think.


<!-- Sent from Reviewable.io -->

DrewKimball reviewed (approved):
This <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> once the test is fixed up. If you can't get the optimizer to choose the right plan, I'm fine with skipping that particular case.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114401#-:-NkvWlww8iNYmOBp5dOM:b-vg3ut3)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @michae2 and @nvanbenschoten)



<!-- Sent from Reviewable.io -->

DrewKimball reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114401#-:-NkxiC7hArfxeicY9wBv:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner, @michae2, and @nvanbenschoten)



<!-- Sent from Reviewable.io -->

michae2 reviewed (commented):
TFTRs!

> If you can't get the optimizer to choose the right plan, I'm fine with skipping that particular case.

I wasn't able to produce a plan with a lock of table `a` on top of an index join of table `b`, so I skipped that for now.

bors r=DrewKimball

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/114401#-:-Nkxl5so33QsRrcON_u3:bhga1kp)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball, @mgartner, and @nvanbenschoten)

___
*[`pkg/sql/opt/exec/execbuilder/mutation.go` line 1173 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/114401#-NjP73Nf5LborIu7vbcC:-NkxlWxvAS4XBI8ZxWtj:b-1dddbr) ([raw file](https://github.com/cockroachdb/cockroach/blob/3c06bdef879e19a36165dd427cdecfdfb759100d/pkg/sql/opt/exec/execbuilder/mutation.go#L1173)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Is it safe to push the locks into locality-optimized lookup joins (e.g., when `RemoteLookupExpr FiltersExpr` is non-nil, or when `LocalityOptimized bool` is true)? I believe we are free to not lock rows if we can answer the query without scanning them, so I think not locking remote rows in a LOS when we find all the rows locally should be ok, correct?
</blockquote></details>

Yes, I think not locking remote rows in a LOS is totally fine. It's similar to a SFU with a LIMIT: we don't need to (and don't want to) lock more rows after hitting the limit.
___
*[`pkg/sql/opt/exec/execbuilder/testdata/select_for_update` line 2773 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/114401#-NjsNWvt2LyWSqA8wMSn:-NkxluIE31H2wpAhg4dd:bwwkji0) ([raw file](https://github.com/cockroachdb/cockroach/blob/d0ea8ec5414592832febfcabb156ca11ef34d251/pkg/sql/opt/exec/execbuilder/testdata/select_for_update#L2773)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Hinting the join should make this easier, I think.
</blockquote></details>

We don't support `INNER INDEX JOIN` as a hint (not sure if this is what you meant).


<!-- Sent from Reviewable.io -->

mgartner commented:
blathers backport 23.2.0-rc
